File: Demo_pageLinkage/MCPageViewController/MCPageViewController.m
//
//  MCPageViewController.m
//  Demo_pageViewController管理多页面
//
//  Created by goulela on 2017/8/17.
//  Copyright © 2017年 MC. All rights reserved.
//

#import "MCPageViewController.h"


@interface MCPageViewController ()
<
UIScrollViewDelegate,UIPageViewControllerDelegate,UIPageViewControllerDataSource
>
{
    NSInteger _curPage;
}
@property (nonatomic, strong) UIScrollView * titleScrollView;
//标题滚动视图
@property (nonatomic, strong) UIView * indicatorView;
//分页控制器
@property (nonatomic, strong) UIPageViewController * pageVC;

/**
 *  标题按钮的数组
 *  用来改变按钮的状态
 */
@property (nonatomic, strong) NSMutableArray * titleButtonArrayM;

@end

@implementation MCPageViewController

#define kReuseCell @"cell"
#define kWidth          self.view.bounds.size.width
#define kHeigth         self.view.bounds.size.height


- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self baseSetting];
    
    [self initUI];
}

#pragma mark - 系统代理
- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController {
    NSInteger index = [_vcArray indexOfObject:viewController];
    if (index == _vcArray.count -1) {
        return nil;
    }
    return _vcArray[index+1];
}
//返回上一页
- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController {
    NSInteger index = [_vcArray indexOfObject:viewController];
    if (index == 0) {
        return nil;
    }
    return _vcArray[index-1];
}

- (void)pageViewController:(UIPageViewController *)pageViewController didFinishAnimating:(BOOL)finished previousViewControllers:(NSArray *)previousViewControllers transitionCompleted:(BOOL)completed {
    
    UIViewController *sub = pageViewController.viewControllers[0];
    NSInteger index = 0;
    for (UIViewController *VC in _vcArray) {
        
        if ([VC isEqual:sub]) {
            _curPage = index;
        }
        index++;
    }
    
    UIButton * btn = (UIButton *)[self.view viewWithTag:_curPage + 1000];
    [self orderButtonClicked:btn];
    _curPage = [_vcArray indexOfObject:_pageVC.viewControllers[0]];

    [self setScrollViewOffSet:btn];
}


//设置偏移
- (void)setScrollViewOffSet:(UIButton *)sender
{
    
    //
    int count = kWidth * 0.5 / self.blockWidth;
    
    if (count % 2 == 0) {
        count --;
    }
    
    CGFloat offsetX = sender.frame.origin.x - count * self.blockWidth;
    if (offsetX<0) {
        offsetX=0;
    }
    CGFloat maxOffsetX= _titleScrollView.contentSize.width - kWidth;
    if (offsetX > maxOffsetX) {
        offsetX = maxOffsetX;
    }
    [UIView animateWithDuration:.2 animations:^{
        [_titleScrollView setContentOffset:CGPointMake(offsetX, 0) animated:YES];
    }];
}


#pragma mark - 点击事件
- (void)orderButtonClicked:(id)sender {
    NSInteger tagNum = [sender tag];
    
    _curPage = tagNum - 1000;
    
    if (_curPage < 0) {
        _curPage = 1;
    }
    
    for (UIButton * button in self.titleButtonArrayM) {
        if (tagNum != button.tag) {
            [button setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        } else {
            [UIView animateKeyframesWithDuration:0.12
                                           delay:0.0
                                         options:UIViewKeyframeAnimationOptionLayoutSubviews
                                      animations:^{
                                          _indicatorView.center = CGPointMake(button.center.x, _indicatorView.center.y);
                                      }
                                      completion:^(BOOL finished) {
                                          [button setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
                                      }];
        }
    }
    [_pageVC setViewControllers:@[_vcArray[_curPage]] direction:_curPage>tagNum animated:YES completion:^(BOOL finished) {
        _curPage = tagNum;
    }];
}

- (void)jumpToVC:(UIButton *)btn {
    //要跳转到的vc索引
    //direction：0代表前进，1代表后退
    [self orderButtonClicked:btn];
    
    NSInteger toPage = btn.tag - 1000;
    [_pageVC setViewControllers:@[_vcArray[toPage]] direction:_curPage>toPage animated:YES completion:^(BOOL finished) {
        _curPage = toPage;
    }];
}

#pragma mark - 实现方法
- (void)baseSetting {

    
    self.view.backgroundColor = [UIColor whiteColor];
    
    if (_barColor == nil) {
        _barColor = [UIColor whiteColor];
    }
    
    if (_barHeight < 40) {
        _barHeight = 40;
    }
    
    if (_blockWidth < 50) {
        _blockWidth = 50;
    }
    
    if (_blockFont < 14) {
        _blockFont = 18;
    }
    
    if (_blockColor == nil) {
        _blockColor = [UIColor whiteColor];
    }
    
    if (_blockNormalColor == nil) {
        _blockNormalColor = [UIColor grayColor];
    }
    
    if (_blockSelectedColor == nil) {
        _blockSelectedColor = [UIColor redColor];
    }
    
    if (_currentPage < 0) {
        _currentPage = 0;
    }
}

- (void)initUI {
    self.titleScrollView.contentSize = CGSizeMake(self.blockWidth*self.titleArray.count, 0);
    self.titleScrollView.frame = CGRectMake(0, 0, kWidth, self.barHeight);
    [self.view addSubview:self.titleScrollView];
    
    //创建按钮
    for (int i = 0; i<self.titleArray.count; i++) {
        UIButton *btn = [UIButton buttonWithType:UIButtonTypeCustom];
        btn.frame = CGRectMake(i*self.blockWidth, 0, self.blockWidth, self.barHeight);
        btn.backgroundColor = self.blockColor;
        btn.titleLabel.font = [UIFont systemFontOfSize:self.blockFont];
        [btn setTitle:[NSString stringWithFormat:@"%@",self.titleArray[i]] forState:UIControlStateNormal];
        [btn setTitleColor:self.blockNormalColor forState:UIControlStateNormal];
        [btn setTitleColor:self.blockSelectedColor forState:UIControlStateNormal];
        btn.tag = 1000 + i;
        //添加点击事件
        [btn addTarget:self action:@selector(jumpToVC:) forControlEvents:UIControlEventTouchUpInside];
        [self.titleButtonArrayM addObject:btn];
        [_titleScrollView addSubview:btn];
    }
    
    [self orderButtonClicked:self.titleButtonArrayM[0]];
    
    self.indicatorView.frame = CGRectMake(0, self.barHeight-1.5, self.blockWidth, 1.5);
    [self.titleScrollView addSubview:self.indicatorView];
    
    self.pageVC.view.frame = CGRectMake(0, self.barHeight, kWidth, kHeigth - self.barHeight);
    [self.view addSubview:self.pageVC.view];
}




#pragma mark - setter & getter
- (UIScrollView *)titleScrollView {
    if (_titleScrollView == nil) {
        self.titleScrollView = [[UIScrollView alloc] init];
        self.titleScrollView.showsHorizontalScrollIndicator = NO;
        self.titleScrollView.backgroundColor = self.barColor;
    } return _titleScrollView;
}

- (UIView *)indicatorView {
    if (_indicatorView == nil) {
        self.indicatorView = [[UIView alloc] init];
        self.indicatorView.backgroundColor = self.blockSelectedColor;
    } return _indicatorView;
}

- (NSMutableArray *)titleButtonArrayM {
    if (_titleButtonArrayM == nil) {
        self.titleButtonArrayM = [NSMutableArray arrayWithCapacity:0];
    } return _titleButtonArrayM;
}

- (UIPageViewController *)pageVC {
    if (!_pageVC) {
        self.pageVC = [[UIPageViewController alloc] initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal options:nil];
        
        self.pageVC.delegate = self;
        self.pageVC.dataSource = self;
        
        [self.pageVC setViewControllers:@[self.vcArray[0]] direction:UIPageViewControllerNavigationDirectionReverse animated:NO completion:nil];
        
        // 遍历pageVC.view的子视图,找到scrollView.设置代理
        for (UIView * view in self.pageVC.view.subviews) {
            if ([view isKindOfClass:[UIScrollView class]]) {
                [view setValue:self forKey:@"delegate"];
            }
        }
    } return _pageVC;
}


@end

File: MCPageViewController/Classes/MCPageViewController/MCPageViewController.swift
//
//  MCPageViewController.swift
//  MCPageViewController
//
//  Created by MC on 2018/11/2.
//  Copyright © 2018年 MC. All rights reserved.
//

import UIKit


public protocol MCPageViewControllerDelegate: NSObjectProtocol {
    
    func pageViewControllerWillBeginDragging(_ pageViewController: MCPageViewController)
    func pageViewControllerDidEndDragging(_ pageViewController: MCPageViewController)
    func pageViewControllerClickMoreEvent(_ pageViewController: MCPageViewController)
}


open class MCPageViewController: UIViewController {
    
    
    public var currentChildPageViewController: MCPageChildViewController?
    
    public weak var delegate: MCPageViewControllerDelegate?
    
    /**
     * 根据数据直接创建
     */
    public func initPagesWithConfig(_ config: MCPageConfig) {
        
        categoryView.categoryModels = config.categoryModels

        settingWithConfig(config)
    }
    
    
    open override func viewDidLoad() {
        super.viewDidLoad()
        
        initBaseUI()
    }
    
    
    
    // MARK: - Setter & Getter
    private lazy var categoryView: MCCategoryBar = {
        let view = MCCategoryBar()
        view.changeItemWithTargetIndex(targetIndex: 0)
        view.delegate = self
        return view
    }()
    
    private lazy var pageVC: UIPageViewController = {
        let vc = UIPageViewController.init(transitionStyle: .scroll, navigationOrientation: .horizontal, options: nil)
        vc.delegate = self
        vc.dataSource = self
        return vc
    }()
}




//MARK: UI的处理,通知的接收
private extension MCPageViewController {
    
    private func settingWithConfig(_ config: MCPageConfig) {
       
        
        
        if (config.categoryModels.count != config.viewControllers.count) || config.categoryModels.count == 0 || config.viewControllers.count == 0 {
            print("MCPageViewController: 请检查config的配置 config.vcs.count:\(config.categoryModels.count) --- config.vcs.count:\(config.viewControllers.count)")
            return
        }
        
        
        

        let index = MCPageConfig.shared.selectIndex
        
        // 指定当前的子控制器
        currentChildPageViewController = config.viewControllers[index] as? MCPageChildViewController
        
        /// 指定当前选中的tab 和  当前选中的viewController
        changeSelectCategoryView(selectIndex: index)
        changeSelectPageViewController(selectPage: index)
    }
    
    
    
    // 基础UI的设置
    private func initBaseUI() {
        
        
        self.view.backgroundColor = UIColor.white

        
        self.view.addSubview(categoryView)
        categoryView.snp.remakeConstraints { (make) ->Void in
            make.left.right.top.equalTo(view)
            make.height.equalTo(50)
        }
        
        
        view.addSubview(pageVC.view)
        pageVC.view.snp.remakeConstraints { (make) ->Void in
            make.left.right.bottom.equalTo(view)
            make.top.equalTo(categoryView.snp.bottom)
        }

    }
}


extension MCPageViewController : UIScrollViewDelegate,UIPageViewControllerDelegate,UIPageViewControllerDataSource {
    
    public func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
        
        let viewControllers = MCPageConfig.shared.viewControllers
        
        
        let index = viewControllers.index(of: viewController) ?? 0
        if index == viewControllers.count - 1 {
            return nil
        }
        return viewControllers[index + 1]
    }
    
    public func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
        
        
        let viewControllers = MCPageConfig.shared.viewControllers

        
        let index = viewControllers.index(of: viewController) ?? 0
        if index == 0 {
            return nil
        }
        
        
        if let index = viewControllers.index(of: viewController) {
            return viewControllers[index - 1]
        } else {
            return nil
        }
    }

    public func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
        
        let sub = pageViewController.viewControllers![0]
        var index = 0
        for subVc in MCPageConfig.shared.viewControllers {
            if subVc.isEqual(sub) { MCPageConfig.shared.selectIndex = index }
            index += 1
        }
        
        /// 滑动了pageViewController，选中对应的分类tab
        changeSelectCategoryView(selectIndex: MCPageConfig.shared.selectIndex)
        currentChildPageViewController = MCPageConfig.shared.viewControllers[MCPageConfig.shared.selectIndex] as? MCPageChildViewController
    }
    
    
    
    public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        delegate?.pageViewControllerWillBeginDragging(self)
    }
    
    public func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        delegate?.pageViewControllerDidEndDragging(self)
    }
    
}

extension MCPageViewController {
    
    /// 更改选中的分类
    func changeSelectCategoryView(selectIndex targetIndex: Int) {
        categoryView.changeItemWithTargetIndex(targetIndex: targetIndex)
    }
    
    /// 更改选中的页面
    func changeSelectPageViewController(selectPage pageIndex: Int)  {
        let toPage = pageIndex
        let direction : UIPageViewController.NavigationDirection = MCPageConfig.shared.selectIndex > toPage ? .forward : .reverse
        
        pageVC.setViewControllers([MCPageConfig.shared.viewControllers[toPage]], direction: direction, animated: false) { (finished) in
            MCPageConfig.shared.selectIndex = toPage;
            
            self.currentChildPageViewController = MCPageConfig.shared.viewControllers[toPage] as? MCPageChildViewController
        }
    }
}

extension MCPageViewController: MCCategoryBarDelegate {
    public func categoryBarClickMoreEvent(categoryBar: MCCategoryBar) {
        delegate?.pageViewControllerClickMoreEvent(self)
    }
    
    
    /// 选中了分类tab，滑动到对应的页面
    public func categoryBar(categoryBar: MCCategoryBar, didSelectItemAt index: Int) {
        changeSelectPageViewController(selectPage: index)
        
        currentChildPageViewController = MCPageConfig.shared.viewControllers[index] as? MCPageChildViewController
        MCPageConfig.shared.selectIndex = index
    }
}
